File: scripts/measure-overhead.mjs
#!/usr/bin/env node
/**
 * Adapter-only operational overhead (#11), JavaScript hosts. Measures and
 * records; it carries no threshold and no verdict. Budgets and judgement
 * belong to redact-secret-benchmarks, which consumes this file's output.
 *
 *   npm run build
 *   node scripts/measure-overhead.mjs --out overhead-js.json
 *   node scripts/measure-overhead.mjs --quick --out -        # CI smoke: shape only, numbers meaningless
 *
 * For every (host, profile) pair it times four modes over the same events,
 * interleaved and rotated per repetition so drift spreads evenly:
 *
 *   host              the host alone (pino, an OpenTelemetry provider), no adapter
 *   adapter-identity  host + adapter over a scanner that finds nothing: traversal and seam cost only
 *   adapter-core      host + adapter over the real `@redact-secret/core`
 *   core-direct       the real core called directly on exactly the leaves the adapter hands it
 *
 * and derives, from per-mode medians, `traversal` (adapter-identity − host),
 * `coreScan` (core-direct), `adapterOverhead` (adapter-core − host), and the
 * `unattributed` remainder. Core scan time is therefore never folded into
 * traversal, and neither is folded into host time.
 */

import { execFileSync } from "node:child_process";
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import os from "node:os";

import { buildEvents, loadProfiles, workloadDigest } from "./overhead-workloads.mjs";

const HOSTS = ["pino", "pino-streamwrite", "otel-js", "mask-js"];
const MODES = ["host", "adapter-identity", "adapter-core", "core-direct"];

function parseArgs(argv) {
  const options = { out: "-", repetitions: 15, events: 400, warmup: 200, hosts: HOSTS, profiles: undefined };
  const setters = {
    "--out": (value) => (options.out = value),
    "--repetitions": (value) => (options.repetitions = Number(value)),
    "--events": (value) => (options.events = Number(value)),
    "--warmup": (value) => (options.warmup = Number(value)),
    "--host": (value) => (options.hosts = value.split(",")),
    "--profile": (value) => (options.profiles = value.split(",")),
  };
  for (let i = 0; i < argv.length; i += 1) {
    const flag = argv[i];
    if (flag === "--quick") {
      Object.assign(options, { repetitions: 3, events: 20, warmup: 5 });
    } else if (setters[flag] !== undefined) {
      setters[flag](argv[i + 1]);
      i += 1;
    } else {
      throw new Error(`unknown argument: ${flag}`);
    }
  }
  for (const key of ["repetitions", "events", "warmup"]) {
    if (!Number.isSafeInteger(options[key]) || options[key] < (key === "warmup" ? 0 : 1)) {
      throw new Error(`--${key} must be a positive integer`);
    }
  }
  return options;
}

/** The installed version of `name`, read from the nearest `node_modules` (ESM-only `exports` hide package.json). */
function versionOf(name) {
  for (let dir = new URL("../", import.meta.url); ; dir = new URL("../", dir)) {
    const manifest = new URL(`node_modules/${name}/package.json`, dir);
    if (existsSync(manifest)) return JSON.parse(readFileSync(manifest, "utf-8")).version;
    if (dir.pathname === "/") return null;
  }
}

function gitState() {
  try {
    const commit = execFileSync("git", ["rev-parse", "HEAD"], { encoding: "utf-8" }).trim();
    const dirty = execFileSync("git", ["status", "--porcelain"], { encoding: "utf-8" }).trim() !== "";
    return { commit, dirty };
  } catch {
    return { commit: null, dirty: null };
  }
}

/** Nearest-rank percentile over sorted samples. */
function percentile(sorted, p) {
  return sorted[Math.max(0, Math.ceil(p * sorted.length) - 1)];
}

function summarize(samples) {
  const sorted = [...samples].sort((a, b) => a - b);
  const mean = samples.reduce((s, x) => s + x, 0) / samples.length;
  const sd = Math.sqrt(samples.reduce((s, x) => s + (x - mean) ** 2, 0) / samples.length);
  const round = (x) => Math.round(x * 1000) / 1000;
  return {
    unit: "microseconds-per-event",
    samples: samples.map(round),
    median: round(percentile(sorted, 0.5)),
    p95: round(percentile(sorted, 0.95)),
    minimum: round(sorted[0]),
    maximum: round(sorted.at(-1)),
    standardDeviation: round(sd),
  };
}

const identityScanner = (text) => ({ text, findings: [] });

async function hostRunners(host) {
  if (host === "pino" || host === "pino-streamwrite") {
    const pino = (await import("pino")).default;
    const { createRedactingLogMethodWith, createRedactingStreamWriteWith } = await import(
      "@redact-secret/adapter-pino"
    );
    const destination = { write() {} };
    const make = (scanner) => {
      const hooks = scanner === undefined ? {} : { logMethod: createRedactingLogMethodWith(scanner) };
      if (scanner !== undefined && host === "pino-streamwrite")
        hooks.streamWrite = createRedactingStreamWriteWith(scanner);
      const logger = pino({ base: null, timestamp: false, hooks }, destination);
      return (event) => logger.info(event.fields, event.message, ...event.args);
    };
    return make;
  }
  if (host === "otel-js") {
    const { BasicTracerProvider, SimpleSpanProcessor } = await import("@opentelemetry/sdk-trace-base");
    const { RedactingSpanProcessorWith } = await import("@redact-secret/adapter-otel");
    const exporter = { export: (_spans, done) => done({ code: 0 }), shutdown: () => Promise.resolve() };
    return (scanner) => {
      const simple = new SimpleSpanProcessor(exporter);
      const processor = scanner === undefined ? simple : new RedactingSpanProcessorWith(simple, scanner);
      const tracer = new BasicTracerProvider({ spanProcessors: [processor] }).getTracer("overhead");
      return (event) => {
        const span = tracer.startSpan(event.name);
        span.setAttributes(event.attributes);
        for (const e of event.events) span.addEvent(e.name, e.attributes);
        span.end();
      };
    };
  }
  if (host === "mask-js") {
    const { maskSecretsWith } = await import("@redact-secret/adapter");
    // There is no host around a masking callback: the baseline is the empty call.
    return (scanner) => (scanner === undefined ? () => undefined : (event) => maskSecretsWith(scanner, event));
  }
  throw new Error(`unknown host: ${host}`);
}

async function measurePair(host, profile, events, core, options) {
  const make = await hostRunners(host);
  // The leaves the adapter hands the core, per event, recorded once.
  const leaves = events.map((event) => {
    const recorded = [];
    make((text) => {
      recorded.push(text);
      return identityScanner(text);
    })(event);
    return recorded;
  });
  const runners = {
    host: make(undefined),
    "adapter-identity": make(identityScanner),
    "adapter-core": make(core.scanAndRedact),
    "core-direct": (() => {
      let cursor = 0;
      return () => {
        for (const text of leaves[cursor]) core.scanAndRedact(text);
        cursor = (cursor + 1) % leaves.length;
      };
    })(),
  };
  // core-direct keeps its own cursor so it walks the same event order as the others.
  const run = (mode, count) => {
    const runner = runners[mode];
    for (let k = 0; k < count; k += 1) runner(events[k % events.length]);
  };
  for (const mode of MODES) run(mode, options.warmup);
  const samples = Object.fromEntries(MODES.map((mode) => [mode, []]));
  for (let rep = 0; rep < options.repetitions; rep += 1) {
    for (let m = 0; m < MODES.length; m += 1) {
      const mode = MODES[(m + rep) % MODES.length];
      const started = process.hrtime.bigint();
      run(mode, options.events);
      samples[mode].push(Number(process.hrtime.bigint() - started) / 1000 / options.events);
    }
  }
  const modes = Object.fromEntries(MODES.map((mode) => [mode, summarize(samples[mode])]));
  const median = (mode) => modes[mode].median;
  const round = (x) => Math.round(x * 1000) / 1000;
  const scannerCalls = leaves.reduce((s, l) => s + l.length, 0) / leaves.length;
  const leafChars = leaves.reduce((s, l) => s + l.reduce((t, x) => t + x.length, 0), 0) / leaves.length;
  return {
    host,
    profileId: profile.id,
    scannerCallsPerEvent: round(scannerCalls),
    scannedCodeUnitsPerEvent: round(leafChars),
    modes,
    derived: {
      unit: "microseconds-per-event",
      basis: "difference of per-mode medians",
      traversal: round(median("adapter-identity") - median("host")),
      coreScan: median("core-direct"),
      adapterOverhead: round(median("adapter-core") - median("host")),
      unattributed: round(median("adapter-core") - median("adapter-identity") - median("core-direct")),
    },
  };
}

async function main() {
  const options = parseArgs(process.argv.slice(2));
  const loadAtStart = os.loadavg();
  const document = loadProfiles();
  const core = await import("@redact-secret/core");
  await core.initialize();
  const results = [];
  for (const profile of document.profiles) {
    if (options.profiles !== undefined && !options.profiles.includes(profile.id)) continue;
    const events = buildEvents(document, profile);
    for (const host of profile.hosts.filter((h) => options.hosts.includes(h) && HOSTS.includes(h))) {
      const result = await measurePair(host, profile, events, core, options);
      results.push(result);
      const d = result.derived;
      console.error(
        `${host}/${profile.id}: host ${result.modes.host.median}µs, traversal ${d.traversal}µs, core ${d.coreScan}µs, overhead ${d.adapterOverhead}µs per event`,
      );
    }
  }
  const cpus = os.cpus();
  const output = {
    schema: "redact-secret-adapters/overhead-v1",
    language: "javascript",
    measuredAt: new Date().toISOString(),
    source: { repository: "redact-secret/redact-secret-adapters", ...gitState() },
    workloads: {
      file: "fixtures/overhead-profiles.json",
      schemaVersion: document.schemaVersion,
      digest: workloadDigest(document),
    },
    environment: {
      os: `${os.platform()}-${os.release()}`,
      platform: os.platform(),
      arch: os.arch(),
      cpuModel: cpus[0]?.model ?? null,
      logicalCpus: cpus.length,
      totalMemoryBytes: os.totalmem(),
      runtime: `node-${process.versions.node}`,
      // Other work on the machine is the main source of noise between runs.
      loadAverage1m: { start: loadAtStart[0], end: os.loadavg()[0] },
      packages: Object.fromEntries(
        [
          "@redact-secret/core",
          "@redact-secret/adapter",
          "@redact-secret/adapter-pino",
          "@redact-secret/adapter-otel",
          "pino",
          "@opentelemetry/sdk-trace-base",
        ].map((name) => [name, versionOf(name)]),
      ),
    },
    method: {
      quick: process.argv.includes("--quick"),
      repetitions: options.repetitions,
      eventsPerRepetition: options.events,
      warmupEventsPerMode: options.warmup,
      distinctEvents: document.distinctEvents,
      order: "modes interleaved within each repetition, rotated by one position per repetition",
      clock: "process.hrtime.bigint",
      percentile: "nearest-rank",
      processes: 1,
    },
    results,
    limitations: [
      "Host-dependent: these numbers describe this machine and runtime only.",
      "Per-event times are batch means over eventsPerRepetition events; the distribution is across repetitions, not across single events.",
      "derived values are differences of medians, not medians of differences, and can be negative within noise.",
      "The OpenTelemetry exporter is a no-op, so export cost is excluded; pino writes to a no-op destination, so I/O cost is excluded.",
      "This output carries no threshold and no verdict.",
    ],
  };
  const text = `${JSON.stringify(output, null, 2)}\n`;
  if (options.out === "-") process.stdout.write(text);
  else writeFileSync(options.out, text);
}

await main();

File: scripts/measure-overhead.py
#!/usr/bin/env python3
"""Adapter-only operational overhead (#11), Python hosts. Measures and
records; it carries no threshold and no verdict. Budgets and judgement
belong to redact-secret-benchmarks, which consumes this file's output.

    pip install -e "./python[otel]"
    python scripts/measure-overhead.py --out overhead-python.json
    python scripts/measure-overhead.py --quick --out -     # CI smoke: shape only, numbers meaningless

Same method as ``scripts/measure-overhead.mjs``: for every (host, profile)
pair it times four modes over the same events, interleaved and rotated per
repetition:

    host              the host alone (a logging handler, an OpenTelemetry provider), no adapter
    adapter-identity  host + adapter over a scanner that finds nothing: traversal and seam cost only
    adapter-core      host + adapter over the real ``redact_secret``
    core-direct       the real core called directly on exactly the leaves the adapter hands it
"""

from __future__ import annotations

import argparse
import importlib.metadata
import json
import logging
import math
import os
import platform
import statistics
import subprocess
import sys
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable, Optional

sys.path.insert(0, str(Path(__file__).resolve().parent))
from overhead_workloads import build_events, load_profiles, workload_digest  # noqa: E402

HOSTS = ("python-logging", "otel-python", "mask-python")
MODES = ("host", "adapter-identity", "adapter-core", "core-direct")


class _Identity:
    __slots__ = ("text", "findings")

    def __init__(self, text: str) -> None:
        self.text = text
        self.findings: list = []


def identity_scanner(text: str, policy: Any = None) -> _Identity:
    return _Identity(text)


class _NullStream:
    def write(self, _text: str) -> int:
        return 0

    def flush(self) -> None:
        pass


def _logging_runner(scanner: Optional[Callable[..., Any]]) -> Callable[[dict[str, Any]], None]:
    from redact_secret_adapters.logging_filter import RedactSecretFilter

    handler = logging.StreamHandler(_NullStream())
    handler.setFormatter(logging.Formatter("%(levelname)s %(message)s %(fields)s"))
    if scanner is not None:
        handler.addFilter(RedactSecretFilter(scanner, extra_fields=("fields",)))
    logger = logging.Logger(f"overhead-{id(handler)}", logging.INFO)
    logger.addHandler(handler)
    logger.propagate = False

    def run(event: dict[str, Any]) -> None:
        logger.info(event["message"], *event["args"], extra={"fields": event["fields"]})

    return run


def _otel_runner(scanner: Optional[Callable[..., Any]]) -> Callable[[dict[str, Any]], None]:
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import SimpleSpanProcessor, SpanExporter, SpanExportResult
    from redact_secret_adapters.otel import RedactingSpanProcessorWith

    class _NullExporter(SpanExporter):
        def export(self, spans):
            return SpanExportResult.SUCCESS

        def shutdown(self) -> None:
            pass

    simple = SimpleSpanProcessor(_NullExporter())
    provider = TracerProvider()
    provider.add_span_processor(simple if scanner is None else RedactingSpanProcessorWith(simple, scanner))
    tracer = provider.get_tracer("overhead")

    def run(event: dict[str, Any]) -> None:
        span = tracer.start_span(event["name"])
        span.set_attributes(event["attributes"])
        for item in event["events"]:
            span.add_event(item["name"], item["attributes"])
        span.end()

    return run


def _mask_runner(scanner: Optional[Callable[..., Any]]) -> Callable[[dict[str, Any]], None]:
    from redact_secret_adapters.mask_secrets import mask_secrets_with

    if scanner is None:
        # There is no host around a masking callback: the baseline is the empty call.
        return lambda _event: None
    return lambda event: mask_secrets_with(scanner, event)


RUNNERS = {"python-logging": _logging_runner, "otel-python": _otel_runner, "mask-python": _mask_runner}


def _percentile(ordered: list[float], p: float) -> float:
    """Nearest-rank percentile over sorted samples."""
    return ordered[max(0, math.ceil(p * len(ordered)) - 1)]


def _summarize(samples: list[float]) -> dict[str, Any]:
    ordered = sorted(samples)
    r = lambda x: round(x, 3)  # noqa: E731
    return {
        "unit": "microseconds-per-event",
        "samples": [r(x) for x in samples],
        "median": r(_percentile(ordered, 0.5)),
        "p95": r(_percentile(ordered, 0.95)),
        "minimum": r(ordered[0]),
        "maximum": r(ordered[-1]),
        "standardDeviation": r(statistics.pstdev(samples)),
    }


def measure_pair(host: str, profile: dict[str, Any], events: list, core: Callable[..., Any], args) -> dict[str, Any]:
    make = RUNNERS[host]
    leaves: list[list[str]] = []
    for event in events:
        recorded: list[str] = []

        def recording(text: str, policy: Any = None, _into: list = recorded) -> _Identity:
            _into.append(text)
            return _Identity(text)

        make(recording)(event)
        leaves.append(recorded)

    cursor = [0]

    def core_direct(_event: dict[str, Any]) -> None:
        for text in leaves[cursor[0]]:
            core(text)
        cursor[0] = (cursor[0] + 1) % len(leaves)

    runners = {
        "host": make(None),
        "adapter-identity": make(identity_scanner),
        "adapter-core": make(core),
        "core-direct": core_direct,
    }

    def run(mode: str, count: int) -> None:
        runner = runners[mode]
        for k in range(count):
            runner(events[k % len(events)])

    for mode in MODES:
        run(mode, args.warmup)
    samples: dict[str, list[float]] = {mode: [] for mode in MODES}
    for rep in range(args.repetitions):
        for m in range(len(MODES)):
            mode = MODES[(m + rep) % len(MODES)]
            started = time.perf_counter_ns()
            run(mode, args.events)
            samples[mode].append((time.perf_counter_ns() - started) / 1000 / args.events)

    modes = {mode: _summarize(samples[mode]) for mode in MODES}
    median = lambda mode: modes[mode]["median"]  # noqa: E731
    return {
        "host": host,
        "profileId": profile["id"],
        "scannerCallsPerEvent": round(sum(map(len, leaves)) / len(leaves), 3),
        "scannedCodeUnitsPerEvent": round(sum(sum(len(t) for t in leaf) for leaf in leaves) / len(leaves), 3),
        "modes": modes,
        "derived": {
            "unit": "microseconds-per-event",
            "basis": "difference of per-mode medians",
            "traversal": round(median("adapter-identity") - median("host"), 3),
            "coreScan": median("core-direct"),
            "adapterOverhead": round(median("adapter-core") - median("host"), 3),
            "unattributed": round(median("adapter-core") - median("adapter-identity") - median("core-direct"), 3),
        },
    }


def _git_state() -> dict[str, Any]:
    try:
        commit = subprocess.check_output(["git", "rev-parse", "HEAD"], text=True).strip()
        dirty = subprocess.check_output(["git", "status", "--porcelain"], text=True).strip() != ""
        return {"commit": commit, "dirty": dirty}
    except (OSError, subprocess.CalledProcessError):
        return {"commit": None, "dirty": None}


def _version(name: str) -> Optional[str]:
    try:
        return importlib.metadata.version(name)
    except importlib.metadata.PackageNotFoundError:
        return None


def _cpu_model() -> Optional[str]:
    try:
        if sys.platform == "darwin":
            return subprocess.check_output(["sysctl", "-n", "machdep.cpu.brand_string"], text=True).strip()
        with open("/proc/cpuinfo", encoding="utf-8") as cpuinfo:
            for line in cpuinfo:
                if line.startswith("model name"):
                    return line.split(":", 1)[1].strip()
    except (OSError, subprocess.CalledProcessError):
        pass
    return platform.processor() or None


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("--out", default="-")
    parser.add_argument("--repetitions", type=int, default=15)
    parser.add_argument("--events", type=int, default=400)
    parser.add_argument("--warmup", type=int, default=200)
    parser.add_argument("--host", default=",".join(HOSTS))
    parser.add_argument("--profile", default=None)
    parser.add_argument("--quick", action="store_true")
    args = parser.parse_args()
    if args.quick:
        args.repetitions, args.events, args.warmup = 3, 20, 5
    if args.repetitions < 1 or args.events < 1 or args.warmup < 0:
        parser.error("--repetitions and --events must be positive, --warmup non-negative")

    import redact_secret

    load_at_start = os.getloadavg()[0] if hasattr(os, "getloadavg") else None

    hosts = [host for host in args.host.split(",") if host in HOSTS]
    only = None if args.profile is None else set(args.profile.split(","))
    document = load_profiles()
    results = []
    for profile in document["profiles"]:
        if only is not None and profile["id"] not in only:
            continue
        events = build_events(document, profile)
        for host in (h for h in profile["hosts"] if h in hosts):
            result = measure_pair(host, profile, events, redact_secret.scan_and_redact, args)
            results.append(result)
            d = result["derived"]
            print(
                f"{host}/{profile['id']}: host {result['modes']['host']['median']}µs, traversal {d['traversal']}µs, "
                f"core {d['coreScan']}µs, overhead {d['adapterOverhead']}µs per event",
                file=sys.stderr,
            )

    output = {
        "schema": "redact-secret-adapters/overhead-v1",
        "language": "python",
        "measuredAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "source": {"repository": "redact-secret/redact-secret-adapters", **_git_state()},
        "workloads": {
            "file": "fixtures/overhead-profiles.json",
            "schemaVersion": document["schemaVersion"],
            "digest": workload_digest(document),
        },
        "environment": {
            "os": f"{platform.system().lower()}-{platform.release()}",
            "platform": sys.platform,
            "arch": platform.machine(),
            "cpuModel": _cpu_model(),
            "logicalCpus": os.cpu_count(),
            "runtime": f"{sys.implementation.name}-{platform.python_version()}",
            # Other work on the machine is the main source of noise between runs.
            "loadAverage1m": {
                "start": load_at_start,
                "end": os.getloadavg()[0] if hasattr(os, "getloadavg") else None,
            },
            "packages": {
                name: _version(name) for name in ("redact-secret", "redact-secret-adapters", "opentelemetry-sdk")
            },
        },
        "method": {
            "quick": args.quick,
            "repetitions": args.repetitions,
            "eventsPerRepetition": args.events,
            "warmupEventsPerMode": args.warmup,
            "distinctEvents": document["distinctEvents"],
            "order": "modes interleaved within each repetition, rotated by one position per repetition",
            "clock": "time.perf_counter_ns",
            "percentile": "nearest-rank",
            "processes": 1,
        },
        "results": results,
        "limitations": [
            "Host-dependent: these numbers describe this machine and runtime only.",
            "Per-event times are batch means over eventsPerRepetition events; "
            "the distribution is across repetitions, not across single events.",
            "derived values are differences of medians, not medians of differences, and can be negative within noise.",
            "The OpenTelemetry exporter is a no-op and the logging stream discards writes, "
            "so export and I/O cost are excluded.",
            "This output carries no threshold and no verdict.",
        ],
    }
    text = json.dumps(output, indent=2) + "\n"
    if args.out == "-":
        sys.stdout.write(text)
    else:
        Path(args.out).write_text(text, encoding="utf-8")


if __name__ == "__main__":
    main()
